File: examples/runtimes/rust/src/node.rs
use async_trait::async_trait;
use std::collections::HashMap;

use crate::types::Context;

/// NodeHandler is the core trait that all Blok nodes must implement.
///
/// # Example
///
/// ```rust
/// use async_trait::async_trait;
/// use blok::{NodeHandler, Context};
/// use std::collections::HashMap;
///
/// struct MyNode;
///
/// #[async_trait]
/// impl NodeHandler for MyNode {
///     async fn execute(
///         &self,
///         ctx: &mut Context,
///         config: &HashMap<String, serde_json::Value>,
///     ) -> Result<serde_json::Value, Box<dyn std::error::Error + Send + Sync>> {
///         let name = ctx.request.body.get("name")
///             .and_then(|v| v.as_str())
///             .unwrap_or("World");
///
///         Ok(serde_json::json!({ "message": format!("Hello, {}!", name) }))
///     }
/// }
/// ```
#[async_trait]
pub trait NodeHandler: Send + Sync {
    /// Execute the node logic with the given workflow context and node configuration.
    ///
    /// - `ctx`: Mutable reference to the workflow context. Nodes can read from
    ///   `ctx.request`, write to `ctx.vars` for downstream nodes, etc.
    /// - `config`: Node-specific configuration from the workflow definition.
    ///
    /// Returns the output data as a JSON value on success, or an error.
    async fn execute(
        &self,
        ctx: &mut Context,
        config: &HashMap<String, serde_json::Value>,
    ) -> Result<serde_json::Value, Box<dyn std::error::Error + Send + Sync>>;
}

File: sdks/go/node.go
package blok

// NodeHandler is the interface that all Blok nodes must implement.
//
// Nodes receive the workflow context and node-specific configuration,
// execute their logic, and return data or an error.
//
// Example:
//
//	type MyNode struct{}
//
//	func (n *MyNode) Execute(ctx *Context, config map[string]interface{}) (interface{}, error) {
//	    name := "World"
//	    if body := ctx.Request.BodyMap(); body != nil {
//	        if v, ok := body["name"].(string); ok {
//	            name = v
//	        }
//	    }
//	    return map[string]interface{}{"message": "Hello, " + name + "!"}, nil
//	}
type NodeHandler interface {
	Execute(ctx *Context, config map[string]interface{}) (interface{}, error)
}

// NodeHandlerFunc is an adapter to allow ordinary functions to be used as NodeHandler.
//
// Example:
//
//	registry.Register("greet", blok.NodeHandlerFunc(func(ctx *blok.Context, config map[string]interface{}) (interface{}, error) {
//	    return map[string]string{"message": "Hello!"}, nil
//	}))
type NodeHandlerFunc func(ctx *Context, config map[string]interface{}) (interface{}, error)

// Execute calls the underlying function.
func (f NodeHandlerFunc) Execute(ctx *Context, config map[string]interface{}) (interface{}, error) {
	return f(ctx, config)
}

// ValidatedNodeHandler is a NodeHandler that also declares input/output schemas for validation.
type ValidatedNodeHandler interface {
	NodeHandler

	// InputSchema returns the JSON Schema for validating input (request body).
	// Return nil to skip input validation.
	InputSchema() map[string]interface{}

	// OutputSchema returns the JSON Schema for validating output (result data).
	// Return nil to skip output validation.
	OutputSchema() map[string]interface{}
}
